File: item-list/src/main/java/com/gurunars/item_list/PlainUpdateFetcher.java
package com.gurunars.item_list;

import java.util.ArrayList;
import java.util.List;

class PlainUpdateFetcher<ItemType extends Item> {

    List<Change<ItemType>> get(List<ItemType> sourceList, List<ItemType> targetList, int offset) {
        List<Change<ItemType>> changes = new ArrayList<>();

        for (int i = 0; i < sourceList.size(); i++) {
            ItemType newItem = targetList.get(i);
            int realIndex = offset + i;
            if (!sourceList.get(i).payloadsEqual(newItem)) {
                changes.add(new ChangeUpdate<>(newItem, realIndex, realIndex));
            }
        }

        return changes;
    }


}

File: item-list/src/main/java/com/gurunars/item_list/PlainUpdateFetcher.kt
package com.gurunars.item_list

import java.util.ArrayList

internal class PlainUpdateFetcher<ItemType : Item> {

    operator fun get(sourceList: List<ItemType>, targetList: List<ItemType>, offset: Int): List<Change<ItemType>> {
        val changes = ArrayList<Change<ItemType>>()

        for (i in sourceList.indices) {
            val newItem = targetList[i]
            val realIndex = offset + i
            if (!sourceList[i].payloadsEqual(newItem)) {
                changes.add(ChangeUpdate(newItem, realIndex, realIndex))
            }
        }

        return changes
    }


}
